feat (DOCS-CODE-017): [sign-in] add canonical blazorserver app (#27)
Azure-Samples/ms-identity-docs-code-dotnet: src/sign-in-blazorserver/README.md

description: "This ASP.NET Core 6 Blazor Server application app signs in the user and then makes a request to Microsoft Graph for the user's profile data."
products:
- azure
- azure-active-directory
- ms-graph
urlFragment: ms-identity-docs-code-app-csharp-blazorserver
---
-->

# ASP.NET Core | Blazor Server application | user sign-in, protected web API access (Microsoft Graph) | Microsoft identity platform

<!-- Build badges here
![Build passing.](https://img.shields.io/badge/build-passing-brightgreen.svg) ![Code coverage.](https://img.shields.io/badge/coverage-100%25-brightgreen.svg) ![License.](https://img.shields.io/badge/license-MIT-green.svg)
-->

This ASP.NET Core Blazor Server application application authenticates a user and then makes a request to the Graph API as the authenticated user. The response to the request is presented to the user.

![A screenshot of a Blazor Server application application displaying a response from Microsoft Graph.](./app-signed-in.png)

## Prerequisites

- Azure Active Directory (Azure AD) tenant and the permissions or role required for managing app registrations in the tenant.

> - [.NET 6.0 SDK](https://dotnet.microsoft.com/download/dotnet/6.0)

Reviewer: Does this need the ASP.NET Core 6 SDK instead?  (I'm not fully up to date with what SDKs ship what now)

Author: this SDK will provider the users with the ASP.NET Core runtime and the scaffoldings/project-templates for `web*`. 

Alternatively, if we wanted to streamline this a bit more, we could just mention the runtime instead. It wouldn't provide a full fleshed dev experience through.

Reviewer: Perfect, good to know.  Nothing to change then.

|--------------------------------:|:--------------------------------------------------------------------|:--------------------------------------------------------------------------------|
| **Name**                        | `active-directory-dotnet-blazorserver`                              | Suggested value for this sample. <br/> You can change the app name at any time. |
| **Supported account types**     | **Accounts in this organizational directory only (Single tenant)**  | Suggested value for this sample.                                                |
| **Platform type**               | **Web**                                                             | Required value for this sample                                                  |
| **Redirect URIs**               | `http://localhost:5000/signin-oidc`                                 | Required value for this sample                                                  |

> :information_source: **Bold text** in the tables above matches (or is similar to) a UI element in the Azure portal, while `code formatting` indicates a value you enter into a text box in the Azure portal.

### 2. Open the project in your code editor

> Next, open the _Api.csproj_ project under the the sign-in-blazorserver folder in your code editor.

Author: sure thing @mmacy 

done | addressed from 96cd2c285e6e535cc8f9cffce5fcf80cf1d72e55

### 3. Update code sample in _appsettings.json_ with app registration values

> Finally, set the following values in _appsettings.json_.

Reviewer 2: ```suggestion
In _appsettings.json_, update each variable with values from the app registration you created earlier.
```

"ClientId": "[Enter the Client Id (Application ID obtained from the Azure portal), e.g. ba74781c2-53c2-442a-97c2-3d60re42f403]",
"ClientSecret": "[Copy the client secret added to the app from the Azure portal]",
```

## Run the application

### 1. Run the webapp

1. Execute the following command to get the app up and running:

   ```bash
   dotnet run
   ```

### 2. Signin into the Blazor Server application

1. Once the web app is listening, navigate to https://localhost:5001. Click the Sign In button and complete the login process.

   ![A screenshot of a Blazor Server application application guiding the user to click the "Sign In" button.](./app-signed-in.png)

1. Click Sign Out

   ![A screenshot of a Blazor Server application application guiding the user to click the "Sign Out" button.](./app-signed-out.png)

## About the code

This ASP.NET Core 6 Blazor Server application is created using .NET Blazor Server. The app is requiring the user to be authenticated in Azure Active Directory.

> When users navigates the home page, the application initiates an authentication flow using the Microsoft Authentication Library (MSAL). The user completes this flow in their default web browser. Upon successful authentication, an HTTP GET request to the Microsoft Graph /me endpoint is issued with the user's access token in the HTTP header. The response from the GET request is then displayed to the user.

Reviewer 2: > Upon successful authentication, an HTTP GET request to the Microsoft Graph /me endpoint is issued with the user's access token in the HTTP header.

This might happen under the covers, but I don't see a raw HTTP GET in the code. Instead, I see a `Microsoft.Identity.Web` reference:

```csharp
using var response = await downstreamWebApi.CallWebApiForUserAsync("GraphApi").ConfigureAwait(false);
```

Also, I don't see a mention of M.I.W. anywhere in the README, which seems pretty important to include. Please take a crack at beefing up this section with the sample's relevant code and architecture details.

As a reminder, here's what's in the comment of the README template for the **About the code** section:

```Markdown
## About the code

<!-- REQUIRED: Short description (3-5 sentences) of the application's -->
<!--           architecture and the code that enables it to function. -->

  <!-- DO discuss ARCHITECTURE and CODE, *not* just what the app *does* (it's features).          -->
  <!-- DO NOT simply restate or slightly rephrase the README's intro paragraph.                   -->
  <!-- DO mention any authNZ LIBRARY used by the app.                                             -->
  <!-- DO discuss the primary LIBRARY TYPES used; how and WHY were they used?                     -->
  <!-- DO explain things like you were talking to a colleague - a fellow software engineer.       -->

<!-- OPTIONAL: DIAGRAM (image or ASCII) showing application architecture and/or authNZ flow. -->
```

For example, you could say it uses M.I.W., the client init is in _Program.cs_, and the code handling sign-in is in _Index.razor_.

The goal of this section is for you to give people a way to know:

- What the app is doing under the hood
- What the app is using to do it (libraries, types, token grant flows)
- Where in the project all the above happens and, most importantly

..._without them having to sift through the code in the project's files to find out_.

For the same reason we try to write the clean, portable, well-documented, and maintainable code, _real_ goal, of course, is so you never, ever have to hear about it again.

In short, **DOC IT INTO SILENCE**.

Reviewer 2: Thought of another way to say it - let me know if this makes sense...

**Answer the question, "What is it DOING?" rather than "What does it DO?"**

Author: good observations and feedback @mmacy 

done | addressed them mostly from 4f5d9768f64b9f3eae8f2dbdf402c3080c0d341f, c7d7fa3cd94db12e819c58ea667e0d1f2766462a, 729eac4573b6810024e8b00a13974919543a2a10, and 92490b23de73faf6298e0a0c42020eaf76cd9684

In _appsettings.json_, update each variable with values from the app registration you created earlier.

```json
"TenantId": "[Enter 'common', or 'organizations' or the Tenant ID (Obtained from the Azure portal. Select 'Endpoints' from the 'App registrations' blade and use the GUID in any of the URLs), e.g. da41245a5-11b3-996c-00a8-4d99re19f292]",
"ClientId": "[Enter the Client Id (Application ID obtained from the Azure portal), e.g. ba74781c2-53c2-442a-97c2-3d60re42f403]",
"ClientSecret": "[Copy the client secret added to the app from the Azure portal]",
```

## Run the application

### 1. Run the webapp

1. Execute the following command to get the app up and running:

   ```bash
   dotnet run
   ```

### 2. Signin into the Blazor Server application

1. Once the web app is listening, navigate to https://localhost:5001. Click the Sign In button and complete the login process.

   ![A screenshot of a Blazor Server application application guiding the user to click the "Sign In" button.](./app-signed-in.png)

1. Click Sign Out

   ![A screenshot of a Blazor Server application application guiding the user to click the "Sign Out" button.](./app-signed-out.png)

## About the code

> This ASP.NET Core 6 Blazor Server application is created using .NET Blazor Server App template. The app is adding sign-in to protect itself, and as a consequence this is requiring the user to be authenticated in Azure Active Directory.

Reviewer: ```suggestion
This ASP.NET Core 6 Blazor Server application is created using the .NET Blazor Server App template. The app is demonstrating the use of Azure Active Directory authentication for user sign-in.
```

Author: Please let accept this suggestion partially, and let the rest for later content dev editorial. 

```suggestion
This ASP.NET Core 6 Blazor Server application is created using the .NET Blazor Server App template. The app is adding sign-in to protect itself, and as a consequence this is requiring the user to be authenticated in Azure Active Directory.
```

> When this .NET Blazor Server starts and before listening to any http call, it bootstraps the application using a single-surface API provided as part of the **Microsoft Identity Web** and the **Microsoft.Idenitty.Web.UI**. The former is tying ASP.NET Core, its authentication middleware for sign-in, and the [Microsoft Authentication Library (MSAL) for .NET](https://github.com/azuread/microsoft-authentication-library-for-dotnet), while the latter adds the UI components and controllers for user sing-in (enter credentials, update password, multi-factor auth, additional consents, etc) and sign-out. All the details required for authentication are being gathered from a configuration section named `AzureAd` as well as others entries used during the call to a protected API like the scopes. As of authorization, it is using the default policy options. Additionally, special services are injected [specifically for Blazor for re-signing, consent and conditional access purposes](https://github.com/AzureAD/microsoft-identity-web/wiki/Managing-incremental-consent-and-conditional-access#in-blazor-server). The recommended pattern to acquire tokens is firstly attempting to acquire them [non-interactevelly (or silent from cache), and then interactively](https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-authentication-flows#interactive-and-non-interactive-authentication). Therefore, it is required to store the tokens for them to be cached. In this tutorial tokens are being stored in memory.

Reviewer: ```suggestion
When this .NET Blazor Server starts and before listening for any HTTP requests, it bootstraps the application using a single-surface API provided as part of **Microsoft.Identity.Web** and **Microsoft.Idenitty.Web.UI**. The former is sign-in middleware for ASP.NET Core that uses [Microsoft Authentication Library (MSAL) for .NET](https://github.com/azuread/microsoft-authentication-library-for-dotnet), while the latter adds UI components and controllers to facilitate user sign-in, sign-out, and other account experiences. All the details required for authentication are being gathered from a configuration section named `AzureAd` as well as others entries used during the call to a protected API, like the scopes. As for authorization, it is using the default policy options. Additionally, special services are injected [specifically for Blazor for re-signing, consent and conditional access purposes](https://github.com/AzureAD/microsoft-identity-web/wiki/Managing-incremental-consent-and-conditional-access#in-blazor-server). The recommended pattern to acquire tokens is first attempting to acquire them [non-interactevelly (or silent from cache), and then interactively](https://docs.microsoft.com/azure/active-directory/develop/msal-authentication-flows#interactive-and-non-interactive-authentication). Therefore, it is required to store the tokens for them to be cached. In this tutorial tokens are being stored in memory.
```

Author: I'm incorporating this, just leaving out a single suggested sentence. Please see [here](https://docs.microsoft.com/en-us/azure/active-directory/develop/scenario-web-app-call-api-overview#specifics) and [here](https://docs.microsoft.com/en-us/azure/active-directory/develop/microsoft-identity-web) for more info.  

After that, If you still consider your sentence was accurate, Im happy to incorporate that one as well. 

```suggestion
When this .NET Blazor Server starts and before listening for any HTTP requests, it bootstraps the application using a single-surface API provided as part of **Microsoft.Identity.Web** and **Microsoft.Idenitty.Web.UI**. The former is tying ASP.NET Core, its authentication middleware for sign-in, and the [Microsoft Authentication Library (MSAL) for .NET](https://github.com/azuread/microsoft-authentication-library-for-dotnet), while the latter adds UI components and controllers to facilitate user sign-in, sign-out, and other account experiences. All the details required for authentication are being gathered from a configuration section named `AzureAd` as well as others entries used during the call to a protected API, like the scopes. As for authorization, it is using the default policy options. Additionally, special services are injected [specifically for Blazor for re-signing, consent and conditional access purposes](https://github.com/AzureAD/microsoft-identity-web/wiki/Managing-incremental-consent-and-conditional-access#in-blazor-server). The recommended pattern to acquire tokens is first attempting to acquire them [non-interactevelly (or silent from cache), and then interactively](https://docs.microsoft.com/azure/active-directory/develop/msal-authentication-flows#interactive-and-non-interactive-authentication). Therefore, it is required to store the tokens for them to be cached. In this tutorial tokens are being stored in memory.
```

> When users navigates the home page, the application initiates an authentication flow, more specifically an **Authorization code type flow (OAuth 2 authorization code grant)**. During this authentication flow, you're being prompted for your credentials if the token has to be acquired interactively, and then asked to consent to the permissions. Upon successful authentication, this web app is making a call to the Microsoft Graph /me endpoint from the **Index** page by using an injected **IDownstreamWebApi** service. This so convenient helper under the covers is making an HTTP GET request to the protected web API adding the cached (if available) or newly acquired user's access token in the HTTP Authorization header. The app displays that you've successfully logged in using your Azure Active Directory credentials, and the Microsft Graph API response.

Reviewer: ```suggestion
When users navigates to the home page, the application initiates an authentication flow, more specifically an **Authorization code type flow (OAuth 2 authorization code grant)**. During this authentication flow, the user is prompted for their credentials, by Azure AD, if the token has to be acquired interactively, and then asked to consent to the permissions. Upon successful authentication, this web app is making a call to the Microsoft Graph /me endpoint by using an injected **IDownstreamWebApi** service. This is facilitates making an HTTP GET request to the protected web API adding the cached or newly acquired user's access token in the HTTP Authorization header. The app displays that you've successfully logged in using your Azure Active Directory credentials, and the Microsoft Graph API response.
```

Author: Accepting mostly all the improvements here just keeping some extra details for now.

```suggestion
When users navigate to the home page, the application initiates an authentication flow, more specifically an **Authorization code type flow (OAuth 2 authorization code grant)**. During this authentication flow, the user is prompted for their credentials, by Azure AD, if the token has to be acquired interactively, and then asked to consent to the permissions. Upon successful authentication, this web app is making a call to the Microsoft Graph /me endpoint from the **Index** page by using an injected **IDownstreamWebApi** service. This helper facilitates making an HTTP GET request to the protected web API adding the cached or newly acquired user's access token in the HTTP Authorization header. The app displays that you've successfully logged in using your Azure Active Directory credentials, and the Microsoft Graph API response.
```

> When a user logs out, their browser is redirected to a public route confirming the user signed-out.

Reviewer 2: Wow! 🥇

## Reporting problems

### Sample app not working?

> If you can't get the sample working, you've checked [Stack Overflow](http://stackoverflow.com/questions/tagged/msal), and you've already searched the issues in this sample's repository, open an issue report the problem.

Reviewer: ```suggestion
If you can't get the sample working, you've checked [Stack Overflow](https://stackoverflow.com/questions/tagged/msal), and you've already searched the issues in this sample's repository, open an issue report the problem.
```
